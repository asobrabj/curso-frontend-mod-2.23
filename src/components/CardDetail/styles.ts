import styled from 'styled-components'
import { Container as Checkbox } from '../Checkbox/styles'

export const Container = styled.div`
  position: relative;
  height: 300px;
  width: 500px;
  padding: 16px;
  border-radius: 16px;
  border: 1px solid ${(props) => props.theme.color.shadow};

  > div {
    margin-bottom: 12px;
  }

  ${Checkbox} {
    margin-top: -8px;
  }
`
export const Paragraph = styled.p`
  margin: 12px 0;
  line-height: 20px;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  font-weight: bold;
`
export const DateHour = styled(Paragraph)`
  color: ${(props) => props.theme.color.shadow};
`

export const ContainerDescription = styled.div`
  margin-top: 16px;
  overflow-wrap: break-word;
  max-width: 100%;
  height: 120px;
  overflow-y: auto;
`

export const Icon = styled.i`
  position: absolute;
  right: 8px;
  top: 5px;
  font-size: 24px;
  cursor: pointer;
  padding: 8px;
  color: ${(props) => props.theme.color.red};
`
