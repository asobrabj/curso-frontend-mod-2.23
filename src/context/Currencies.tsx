import { createContext, useContext, useState } from 'react';
import instance from '../service/axiosaInstace';

export interface ICurrencyContext {
  to: string;
  setTo: React.Dispatch<React.SetStateAction<string>>;
  from: string;
  setFrom: React.Dispatch<React.SetStateAction<string>>;
  amount: string;
  setAmount: React.Dispatch<React.SetStateAction<string>>;
  fecthConversion: (
    to: string,
    from: string,
    amount: string,
  ) => Promise<string>;
  error: boolean;
}

interface ICurrencyProvider {
  children: React.ReactNode;
}

const CurrencyContext = createContext<ICurrencyContext>({} as ICurrencyContext);

export const CurrencyProvider: React.FC<ICurrencyProvider> = ({ children }) => {
  const [error, setError] = useState<boolean>(false);
  const [to, setTo] = useState<string>('');
  const [from, setFrom] = useState<string>('');
  const [amount, setAmount] = useState<string>('');

  const fecthConversion = async (
    to: string,
    from: string,
    amount: string,
  ): Promise<string> => {
    try {
      const response = await instance.get(`${to}/${from}/${amount}`);
      return response.data.conversion_result;
    } catch (err) {
      setError(true);
      return 'Erro na convers√£o';
    }
  };

  return (
    <CurrencyContext.Provider
      value={{
        fecthConversion,
        error,
        amount,
        from,
        to,
        setTo,
        setAmount,
        setFrom,
      }}
    >
      {children}
    </CurrencyContext.Provider>
  );
};

export const useCurrency = () => {
  return useContext(CurrencyContext);
};
