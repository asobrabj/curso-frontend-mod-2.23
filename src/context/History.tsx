/* eslint-disable react-refresh/only-export-components */
import { createContext, useContext, useEffect, useState } from 'react';

export interface IDataCurrency {
  to: string;
  from: string;
  amount: string;
  result: string;
  date: string;
}

export interface IHistoryContext {
  setHistory: (data: IDataCurrency) => void;
  history?: Array<IDataCurrency>;
}

interface IHistoryProvider {
  children: React.ReactNode;
}

const HistoryContext = createContext<IHistoryContext>({} as IHistoryContext);

export const HistoryProvider: React.FC<IHistoryProvider> = ({ children }) => {
  const [history, setHistoryState] = useState<Array<IDataCurrency>>();

  useEffect(() => {
    const currentData = window.localStorage.getItem('history');
    const newData = currentData ? JSON.parse(currentData) : [];
    setHistoryState(newData);
  }, []);

  const setHistory = (data: IDataCurrency) => {
    const currentData = window.localStorage.getItem('history');
    const newData = currentData ? JSON.parse(currentData) : [];
    newData.push(data);
    setHistoryState(newData);
    window.localStorage.setItem('history', JSON.stringify(newData));
    return;
  };

  return (
    <HistoryContext.Provider value={{ setHistory, history }}>
      {children}
    </HistoryContext.Provider>
  );
};

export const useHistory = () => {
  return useContext(HistoryContext);
};
