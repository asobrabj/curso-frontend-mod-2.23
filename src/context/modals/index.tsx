import { createContext, useContext, useState } from 'react'
import IAnotation from '../../types/anotation'

export interface IModalContext {
  isVisible: boolean
  setIsVisible: React.Dispatch<React.SetStateAction<boolean>>
  isForm: boolean
  setIsForm: React.Dispatch<React.SetStateAction<boolean>>
  dataCard: IAnotation
  setDataCard: React.Dispatch<React.SetStateAction<IAnotation>>
  idCard: number | null
  setIdCard: React.Dispatch<React.SetStateAction<number | null>>
}

interface IModalProvider {
  children: React.ReactNode
}

const ModalContext = createContext<IModalContext>({} as IModalContext)

export function ModalProvider({ children }: IModalProvider) {
  const [isVisible, setIsVisible] = useState<boolean>(false)
  const [isForm, setIsForm] = useState<boolean>(false)
  const [idCard, setIdCard] = useState<number | null>(null)
  const [dataCard, setDataCard] = useState<IAnotation>({
    id: null,
    concluded: false,
    date: '',
    description: '',
    name: '',
    category: '',
    priority: '',
    time: '',
  })

  return (
    <ModalContext.Provider
      value={{
        idCard,
        setIdCard,
        isVisible,
        setIsVisible,
        isForm,
        setIsForm,
        dataCard,
        setDataCard,
      }}
    >
      {children}
    </ModalContext.Provider>
  )
}

export function useModal() {
  return useContext(ModalContext)
}
