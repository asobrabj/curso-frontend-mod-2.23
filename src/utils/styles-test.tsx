import { render } from '@testing-library/react'
import { ThemeProvider } from 'styled-components'
import theme from '../styles/darkTheme'

export const renderTheme = (Children: React.ReactNode) => {
  return render(<ThemeProvider theme={theme}> {Children} </ThemeProvider>)
}
