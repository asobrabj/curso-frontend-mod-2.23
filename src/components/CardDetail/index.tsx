import { MdOutlineClose } from 'react-icons/md'
import { useModal } from '../../context/modals'
import dataList from '../../model/list'
import CheckBox from '../Checkbox'
import Tag from '../Tag'
import * as S from './styles'

export default function CardDetail() {
  const { setIsVisible, dataCard } = useModal()
  const [categories, priorities] = dataList

  const handleClose = () => {
    setIsVisible(false)
  }
  const { concluded, date, description, name, time, category, priority, id } =
    dataCard

  const categorySelected = categories.options.find(
    (item) => item.text === category
  )
  const prioritySelected = priorities.options.find(
    (item) => item.text === priority
  )
  return (
    <S.Container>
      <S.Icon onClick={handleClose}>
        <MdOutlineClose />
      </S.Icon>
      {category && (
        <Tag
          background={categorySelected!.background}
          name="categoria"
          text={categorySelected!.text}
        ></Tag>
      )}
      {priority && (
        <Tag
          background={prioritySelected!.background}
          name="prioridade"
          text={prioritySelected!.text}
        ></Tag>
      )}
      <S.DateHour>
        {date}: {time}
      </S.DateHour>
      <CheckBox checked={concluded} id={id as number} />
      <S.Paragraph>{name}</S.Paragraph>
      <S.ContainerDescription>{description}</S.ContainerDescription>
    </S.Container>
  )
}
