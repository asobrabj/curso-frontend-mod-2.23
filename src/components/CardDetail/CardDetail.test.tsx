import { fireEvent, screen, waitFor } from '@testing-library/react'
import { ModalProvider, useModal } from '../../context/modals'
import { renderTheme } from '../../utils/styles-test'
import CardDetail from './'

jest.mock('../../context/modals')

const mockUseModal = useModal as jest.MockedFunction<typeof useModal>

const mockData = {
  id: 1,
  name: 'Test Card',
  date: '2023-12-20',
  time: '00:00',
  category: 'Categoria1',
  priority: 'Alta',
  concluded: false,
  description: 'Teste Jest',
}

describe('CardDetail Component', () => {
  beforeEach(() => {
    jest.clearAllMocks()
  })

  it('deve renderizar corretamente com os dados fornecidos', () => {
    mockUseModal.mockReturnValueOnce({
      setIsVisible: jest.fn(),
      dataCard: mockData,
    })

    renderTheme(
      <ModalProvider>
        <CardDetail />
      </ModalProvider>
    )
    waitFor(() => {
      expect(screen.getByText(mockData.name)).toBeInTheDocument()
      expect(
        screen.getByText(`${mockData.date}: ${mockData.time}`)
      ).toBeInTheDocument()
      expect(screen.getByText(mockData.description)).toBeInTheDocument()
      expect(screen.getByLabelText('checkbox')).toBeInTheDocument()
    })
  })

  it('deve chamar a função setIsVisible ao fechar o card', () => {
    mockUseModal.mockReturnValueOnce({
      setIsVisible: jest.fn(),
      dataCard: mockData,
    })

    renderTheme(
      <ModalProvider>
        <CardDetail />
      </ModalProvider>
    )

    waitFor(() => {
      fireEvent.click(screen.getByLabelText('Fechar'))
      expect(mockUseModal.setIsVisible).toHaveBeenCalledWith(false)
    })
  })
  it('deve exibir a categoria e a prioridade corretamente', () => {
    mockUseModal.mockReturnValueOnce({
      setIsVisible: jest.fn(),
      dataCard: mockData,
    })

    renderTheme(
      <ModalProvider>
        <CardDetail />
      </ModalProvider>
    )
    waitFor(() => {
      expect(
        screen.getByText(`Categoria: ${mockData.category}`)
      ).toBeInTheDocument()
      expect(
        screen.getByText(`Prioridade: ${mockData.priority}`)
      ).toBeInTheDocument()
    })
  })

  it('deve exibir mensagem se não houver descrição', () => {
    const dataWithoutDescription = { ...mockData, description: undefined }

    mockUseModal.mockReturnValueOnce({
      setIsVisible: jest.fn(),
      dataCard: dataWithoutDescription,
    })

    renderTheme(
      <ModalProvider>
        <CardDetail />
      </ModalProvider>
    )
    waitFor(() => {
      expect(
        screen.getByText('Nenhuma descrição disponível.')
      ).toBeInTheDocument()
    })
  })

  it('snapshot', () => {
    const { container } = renderTheme(
      <ModalProvider>
        <CardDetail />
      </ModalProvider>
    )
    expect(container).toMatchSnapshot()
  })
})
