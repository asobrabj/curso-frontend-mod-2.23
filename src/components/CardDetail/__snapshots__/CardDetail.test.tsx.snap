// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardDetail Component snapshot 1`] = `
<div>
   
   
</div>
`;
